core/services/blockhashstore/feeder.go

```diff
@@ -114,3 +132,114 @@ func (f *Feeder) Run(ctx context.Context) error {
 	f.lastRunBlock = latestBlock
 	return errs
 }
+
+func (f *Feeder) runTrusted(
+	ctx context.Context,
+	latestBlock uint64,
+	fromBlock uint64,
+	blockToRequests map[uint64]map[string]struct{},
+) error {
+	var errs error
+
+	// Iterate through each request block via waitGroup.
+	// For blocks with pending requests, add them to the batch to be stored.
+	// Note: Golang maps sort items in a range randomly, so although the batch size is used
+	// to limit blocks-per-batch, every block has an equal chance of getting picked up
+	// on each run.
+	var batch = make(map[uint64]struct{})
+	for blockKey, unfulfilledReqs := range blockToRequests {
+		f.wgStored.Add(1)
+		var unfulfilled = unfulfilledReqs
+		var block = blockKey
+		go func() {
+			defer f.wgStored.Done()
+			if len(unfulfilled) == 0 {
+				return
+			}
+			if _, ok := f.stored[block]; ok {
+				// Already stored
+				return
+			}
+
+			// Do not store a block if it has been marked as stored; otherwise, store it even
+			// if the RPC call errors, as to be conservative.
+			stored, err := f.bhs.IsStored(ctx, block)
+			if err != nil {
+				f.lggr.Errorw("Failed to check if block is already stored, attempting to store anyway",
+					"err", err,
+					"block", block)
+				errs = multierr.Append(errs, errors.Wrap(err, "checking if stored"))
+			} else if stored {
+				f.lggr.Infow("Blockhash already stored",
+					"block", block, "latestBlock", latestBlock,
+					"unfulfilledReqIDs", LimitReqIDs(unfulfilled, 50))
+				f.stored[block] = struct{}{}
+				return
+			}
+
+			// If there's room, store the block in the batch. Threadsafe.
+			f.batchLock.Lock()
+			if len(batch) < int(f.trustedBHSBatchSize) {
+				batch[block] = struct{}{}
+			}
+			f.batchLock.Unlock()
+		}()
+	}
+
+	// Ensure all blocks are checked before storing the batch.
+	f.wgStored.Wait()
+
+	// For a non-empty batch, store all blocks.
+	if len(batch) != 0 {
+		var blocksToStore []uint64
+		var blockhashesToStore []common.Hash
+		var latestBlockhash common.Hash
+
+		// Get all logpoller blocks for the range including the batch and the latest block,
+		// as to include the recent blockhash.
+		lpBlocks, err := f.lp.GetBlocksRange(ctx, append(maps.Keys(batch), latestBlock))
```

Comment:
Is LP guaranteed to have all these blocks in the range or does it prune blocks from it's tables?

Reply:
I'm not sure if they ever agreed to prune blocks from the LP but it definitely shouldn't be within a ~1000 or even ~10,000 block range, which is what we would reasonably use here.